src/docker/ContextManager.ts

```diff
@@ -195,133 +198,200 @@ export class DockerContextManager implements ContextManager, Disposable {
 
-            try {
-                if (isNewContextType(currentContext.ContextType)) {
-                    actionContext.telemetry.properties.hostProtocol = currentContext.ContextType;
-                } else {
-                    actionContext.telemetry.properties.hostProtocol = new URL(currentContext.DockerEndpoint).protocol;
-                }
-            } catch {
-                actionContext.telemetry.properties.hostProtocol = 'unknown';
-            }
+            return dockerContext;
+        }
 
-            if (dockerContextEnv) {
-                this.setVsCodeContext('vscode-docker:contextLocked', true);
-            } else {
-                this.setVsCodeContext('vscode-docker:contextLocked', false);
-            }
+        return undefined;
+    }
 
-            return result;
-        });
+    private async tryGetContextFromEnvironment(actionContext: IActionContext): Promise<DockerContext | undefined | string> {
+        let dockerHost: string | undefined;
+        let dockerContext: string | undefined;
 
-        // If the load failed or is otherwise empty, return the default
-        // That way a returned value is ensured by this method
-        if (!loadResult) {
-            loadResult = [{
+        if ((dockerHost = process.env.DOCKER_HOST)) { // Assignment + check is intentional
+            actionContext.telemetry.properties.hostSource = 'env';
+
+            return {
+                ...defaultContext,
+                Current: true,
+                DockerEndpoint: dockerHost,
+            } as DockerContext;
+        } else if ((dockerContext = process.env.DOCKER_CONTEXT)) { // Assignment + check is intentional
+            actionContext.telemetry.properties.hostSource = 'envContext';
+
+            return dockerContext;
+        }
+
+        return undefined;
+    }
+
+    private async tryGetContextFromFilesystemClues(actionContext: IActionContext): Promise<DockerContext | undefined> {
+        // If there's nothing inside ~/.docker/contexts/meta (or it doesn't exist), then there's only the default, unmodifiable DOCKER_HOST-based context
+        // It is unnecessary to call `docker context inspect`
+        if (!fse.pathExistsSync(dockerContextsFolder) || fse.readdirSync(dockerContextsFolder).length === 0) { // Sync is intentionally used for performance, this is on the activation code path
+            actionContext.telemetry.properties.hostSource = 'defaultContextOnly';
+
+            return {
                 ...defaultContext,
                 Current: true,
                 DockerEndpoint: isWindows() ? WindowsLocalPipe : UnixLocalPipe,
-            } as DockerContext];
+            } as DockerContext;
         }
 
-        return loadResult;
+        return undefined;
+    }
+
+    private async tryGetContextsFromGrpc(actionContext: IActionContext, maybeFixedContextName: string | undefined): Promise<DockerContext[] | undefined> {

```

Comment:
gRPC is an implementation detail. Suggest `tryGetContextsUsingAPI`, or something along these lines

Reply:
Good idea.